src/experimental/tools/VideoThumbnailLoader/load_segments.ts

```diff
@@ -0,0 +1,63 @@
+import { combineLatest, Observable } from "rxjs";
+import {
+  ISegmentLoaderContent,
+  ISegmentLoaderEvent,
+} from "../../../core/fetchers/segment/create_segment_loader";
+import { ISegment } from "../../../manifest";
+import createRequest, { ICancellableRequest } from "./create_request";
+import { IContentInfos } from "./types";
+
+/**
+ * Load needed segments, from already running requests or by running new ones.
+ * @param {Array.<Object>} segments
+ * @param {Map} currentRequests
+ * @param {Function} segmentLoader
+ * @param {Object} contentInfos
+ * @returns {Object}
+ */
+export default function loadSegments(
+  segments: ISegment[],
+  currentRequests: Map<string, ICancellableRequest>,
+  segmentLoader: (
+    x : ISegmentLoaderContent
+  ) => Observable<ISegmentLoaderEvent<Uint8Array |
+                                      ArrayBuffer |
+                                      null>>,
+  contentInfos: IContentInfos
+): Observable<Array<{ segment: ISegment;
+                      data: Uint8Array; }>> {
+  return combineLatest(
+    segments.map((segment) => {
+      return new Observable<{
+        segment: ISegment;
+        data: Uint8Array; }
+      >((obs) => {
+        let request = currentRequests.get(segment.id);
+        if (request === undefined) {
+          request = createRequest(segmentLoader, contentInfos, segment);
+          currentRequests.set(segment.id, request);
+        }
+
+        if (request.error !== undefined) {
+          obs.error(request.error);
+          return;
+        }
+
+        request.onError = (err) => {
+          obs.error(err);
+        };
+
+        if (request.data !== undefined) {
+          obs.next({ data: request.data, segment });
+          obs.complete();
+          return;
+        }
+
+        request.onData = (data) => {
+          obs.next({ data, segment });
+          obs.complete();
```

Comment:
And here? (And in the two synchronous places in here?)

Reply:
Not sure about here. I think we should remove it when there is an error or when the data has been pushed. A task may be interrupted before pushing to the buffer.